tests/c/udp-push-pop.c

```diff
@@ -0,0 +1,238 @@
+ * Imports                                                                                                            *
+ *====================================================================================================================*/
+
+#include <arpa/inet.h>
+#include <assert.h>
+#include <demi/libos.h>
+#include <demi/sga.h>
+#include <demi/wait.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+
+/*====================================================================================================================*
+ * Constants                                                                                                          *
+ *====================================================================================================================*/
+
+/**
+ * @brief Data size.
+ */
+#define DATA_SIZE 64
+
+/**
+ * @brief Maximum number of iterations.
+ */
+#define MAX_ITERATIONS 1000000
+
+/*====================================================================================================================*
+ * sighandler()                                                                                                       *
+ *====================================================================================================================*/
+
+/**
+ * @brief Signal handler.
+ *
+ * @param signum Number of received signal.
+ */
+static void sighandler(int signum)
+{
+    const char *signame = strsignal(signum);
+
+    fprintf(stderr, "\nReceived %s signal\n", signame);
+    fprintf(stderr, "Exiting...\n");
+
+    exit(EXIT_SUCCESS);
+}
+
+/*====================================================================================================================*
+ * server()                                                                                                           *
+ *====================================================================================================================*/
+
+/**
+ * @brief UDP server.
+ *
+ * @param argc   Argument count.
+ * @param argv   Argument list.
+ * @param local  Local socket address.
+ */
+static void server(int argc, char *const argv[], struct sockaddr_in *local)
+{
+    int sockqd = -1;
+
+    /* Initialize demikernel */
+    assert(demi_init(argc, argv) == 0);
+
+    /* Setup peer. */
+    assert(demi_socket(&sockqd, AF_INET, SOCK_DGRAM, 0) == 0);
+    assert(demi_bind(sockqd, (const struct sockaddr *)local, sizeof(struct sockaddr_in)) == 0);
+
+    /* Run peer. */
+    for (int it = 0; it < MAX_ITERATIONS; it++)
+    {
+        demi_qtoken_t qt = -1;
+        demi_qresult_t qr = {0};
+
+        /* Push data. */
+        assert(demi_pop(&qt, sockqd) == 0);
+
+        /* Wait push operation to complete. */
+        assert(demi_wait(&qr, qt) == 0);
+
+        /* Parse operation result. */
+        assert(qr.qr_opcode == DEMI_OPC_POP);
+        assert(qr.qr_value.sga.sga_segs != 0);
+
```

Comment:
Should you look at the data here?  Depends upon what you're trying to test, I suppose.

Reply:
Sure. I've introduced support for payload checking.

Reply:
Okay.